pkg/kubelet/apis/device-plugin/v1alpha1/api.proto

```diff
@@ -0,0 +1,105 @@
+// To regenerate api.pb.go run hack/update-generated-runtime.sh
+syntax = 'proto3';
+
+package deviceplugin;
+
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+
+option (gogoproto.goproto_stringer_all) = false;
+option (gogoproto.stringer_all) =  true;
+option (gogoproto.goproto_getters_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.sizer_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.goproto_unrecognized_all) = false;
+
+
+// Registration is the service advertised by the Kubelet
+// Only when Kubelet answers with a success code to a Register Request
+// may Device Plugins start their service
+// Registration may fail when device plugin version is not supported by
+// Kubelet or the registered resourceName is already taken by another
+// active device plugin. Device plugin is expected to terminate upon registration failure
+service Registration {
+	rpc Register(RegisterRequest) returns (RegisterResponse) {}
+}
+
+// DevicePlugin is the service advertised by Device Plugins
+service DevicePlugin {
+	// ListWatch returns a stream of List of Devices
+	// Whenever a Device state change or a Device disapears, ListWatch
+	// returns the new list
+	rpc ListWatch(Empty) returns (stream ListWatchResponse) {}
+
+	// Allocate is called during container creation so that the Device
+	// Plugin can run device specific operations and instruct Kubelet
+	// of the steps to make the Device available in the container
+	rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+}
+
+message RegisterRequest {
+	// Version of the API the Device Plugin was built against
+	string version = 1;
+	// Name of the unix socket the device plugin is listening on
+	string endpoint = 2;
+	// Name of the Device advertised
+	string resourceName = 3;
+}
+
+message RegisterResponse {
+	// Minimum version the Kubelet API supports.
+	bool success = 1;
+	// Kubelet fills this field if it encounters any errors
+	// during the registration process
+	string error = 2;
+}
+
+// Allocate is called during container creation so that the Device
+// Plugin can run device specific operations on the Devices requested
+message AllocateRequest {
+	repeated string devicesIDs = 1;
+}
+
+message AllocateResponse {
+	repeated DeviceRuntimeSpec spec = 1;
+}
+
+// ListWatch returns a stream of List of Devices
+// Whenever a Device state change or a Device disapears, ListWatch
+// returns the new list
+message ListWatchResponse {
+	repeated Device devices = 1;
+}
+
+// The list to be added to the CRI spec
+message DeviceRuntimeSpec {
```

Comment:
> Why isn't a device file included? How will kubelet know what device file to expose unless the plugin instructs it?

It's the Mount struct.

> On that note, doesn't the DeviceRuntimeSpec have to be per device ID?

Yes, is your point that we should have the device ID in the RuntimeSpec instead of having the same order as the one given by allocate ?